File: cluster/src/main/java/net/kuujo/copycat/cluster/Cluster.java
/*
 * Copyright 2015 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.kuujo.copycat.cluster;

import net.kuujo.copycat.AbstractResource;
import net.kuujo.copycat.Listener;
import net.kuujo.copycat.ListenerContext;
import net.kuujo.copycat.Listeners;
import net.kuujo.copycat.raft.Raft;

/**
 * Cluster resource.
 *
 * @author <a href="http://github.com/kuujo">Jordan Halterman</a>
 */
public class Cluster extends AbstractResource {
  private final Listeners<Member> joinListeners = new Listeners<>();
  private final Listeners<Member> leaveListeners = new Listeners<>();

  public Cluster(Raft protocol) {
    super(protocol);
  }

  /**
   * Listens for a member joining the cluster.
   *
   * @param listener The member listener.
   * @return The listener context.
   */
  public ListenerContext<Member> onMemberJoined(Listener<Member> listener) {
    return joinListeners.add(listener);
  }

  /**
   * Listens for a member leaving the cluster.
   *
   * @param listener The member listener.
   * @return The listener context.
   */
  public ListenerContext<Member> onMemberLeft(Listener<Member> listener) {
    return leaveListeners.add(listener);
  }

}

File: cli/commands/cluster.py
# -*- coding: utf-8

# Copyright 2017-present Open Networking Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

from . import Command, Action, Resource, command


class NodesAction(Action):
    def execute(self):
        self.cli.service.output(self.cli.service.get(self.cli.service.url('/v1/cluster/nodes')))


class NodeAction(Action):
    def execute(self, node=None):
        if node is None:
            self.cli.service.output(self.cli.service.get(
                self.cli.service.url('/v1/cluster/node')
            ))
        else:
            self.cli.service.output(self.cli.service.get(
                self.cli.service.url('/v1/cluster/nodes/{node}', node=node)
            ))


class NodeResource(Resource):
    def _get_nodes(self):
        response = self.cli.service.get(self.cli.service.url('/v1/cluster/nodes'), log=False)
        if response.status_code == 200:
            return response.json()
        return []

    def suggest(self, prefix):
        for node in self._get_nodes():
            if node['id'].lower().startswith(prefix.lower()):
                return node['id'][len(prefix):]
        return None

    def complete(self, prefix):
        for node in self._get_nodes():
            if node['id'].lower().startswith(prefix.lower()):
                yield node['id']


@command(
    'cluster node [id]',
    type=Command.Type.CLUSTER,
    node=NodeAction,
    id=NodeResource
)
@command(
    'cluster nodes',
    type=Command.Type.CLUSTER,
    nodes=NodesAction
)
class ClusterCommand(Command):
    """Cluster command"""